openedx-unsupported/ecommerce
# Multitenant payment processor config

**Description:** Adds multitenancy support for payment processor configuration, using Django sites framework.
**Dependencies:** #629 
**Merge deadline:** 1 Apr
**Related:** [Discovery document](https://docs.google.com/document/d/17BRYvkO4ETvBV9TFodElpbJaK4oze7e0mNqf9ztnGnA/edit#heading=h.70eeasm5mr1q)
**Sandbox:** 
- edX site: [LMS](http://pr11898.sandbox.opencraft.com), [Studio](http://studio.pr11898.sandbox.opencraft.com), [eCommerce](https://ecommerce.pr11898.sandbox.opencraft.com)
- WL site: [LMS](http://wlsite.pr11898.sandbox.opencraft.com), [eCommerce](https://ecommerce.wlsite.pr11898.sandbox.opencraft.com)

Note both HTTP and HTTPS are available for LMS and Otto.

**JIRA:** [WL-362](https://openedx.atlassian.net/browse/WL-362)

**Preparing environment**
1. Switch to current branch
2. Modify settings - move current contents of PAYMENT_PROCESSOR_CONFIG into a new nested dict with  `default` key, i.e. 
    `PAYMENT_PROCESSOR_CONFIG = { 'default': OLD_PAYMENT_PROCESSOR_CONFIG_VALUE }`
3. Migrate DB (adds "OAuth settings" field to `SiteConfiguration`) model.

**LMS and Otto configuration**

The idea is to set up two Sites in Otto and configure Otto and LMS so that both sites are operational.
We will be using two "sites": "edx" and "wl"; so [sitename] is one of (`edx`,  `wl`)

**Configuring LMS:**

The goal here is to configure two (or more) OAuth clients.
1. Go to http://localhost:8000/admin/oauth2/client/ -> Add Client
2. Set values: 
   1. _Url:_ http://[sitename].lms.local:8002
   2. _Redirect Uri_: http://[sitename].lms.local:8002/complete/edx-oidc/
   3. _Client Id:_ ecommerce-[sitename]-key
   4. _Client Secret:_ ecommerce-[sitename]-secret
   5. _Client type:_ Confidential
3. Go to http://localhost:8000/admin/oauth2_provider/trustedclient/ and designate both clients added in (2) as trusted (add one record per client)
4. In http://localhost:8000/admin/commerce/commerceconfiguration/ add/edit the configuration and set check "Checkout on ecommerce service".
5. Make sure FEATURES['ENABLE_OAUTH2_PROVIDER'] is set to True in lms.env.json (required to generate credit course records in course catalog - can be skipped if no credit-mode courses are used)

**Configuring Otto**

Otto configuration: Configure two Sites and SiteConfiguration records so that Otto can connect to LMS
1. Add two site records at http://localhost:8002/admin/sites/site/ (i.e. edx.lms.local:8002 and wl.lms.local:8002)
2. Add two partner records (http://localhost:8002/admin/partner/partner/) - choose unique name and short_code, don't select users.
3. Add corresponding SiteConfiguration records (http://mitpe.lms.analytics.local:8002/admin/core/siteconfiguration/):
   1. _Site:_ choose one of sites selected earlier
   2. _Parner:_ choose unique partner (or create new)
   3. _Path to custom site theme:_ anything, but empty string (required, but works fine if file does not exist)
   4. *Payment processors:' comma separated list of supported processors (supported so far: paypal, cybersource), i.e. `paypal,cybersource`
   5. OAuth settings:

```
    {
      "SOCIAL_AUTH_EDX_OIDC_ID_TOKEN_DECRYPTION_KEY":"ecommerce-[sitename]-secret",
      "SOCIAL_AUTH_EDX_OIDC_KEY":"ecommerce-[sitename]-key",
      "SOCIAL_AUTH_EDX_OIDC_SECRET":"ecommerce-[sitename]-secret",
      "SOCIAL_AUTH_EDX_OIDC_URL_ROOT": "http://[sitename].lms.local:8000/oauth2",
      "OAUTH_PROVIDER_URL":"http://[sitename].lms.local:8000/"
    }
```

**Set up payment processor configuration in Otto**

Add payment processor configuration for each site - add new top level records to PAYMENT_PROCESSOR_CONFIG setting:

```
PAYMENT_PROCESSOR_CONFIG = {
    'default': { ... },
    '[sitename]': { 
        'cybersource': { $cybersource_config$ },
        'paypal': { $paypal_config$ },
}
```

$cybersource_config$ simplest way to do everything right is to copy PAYMENT_PROCESSOR_CONFIG['default'] value and than replace relevant settings. For cybersource 'merchant_id', 'transaction_key':  'profile_id', 'access_key',  'secret_key' should be replaced.

**Configure WL microsite in LMS**

TBD - this part is reworked by @mattdrayer (microsite -> comprehensive theming). Meanwhile, just continue using microsite framework - add the following to lms.env.json:

```
"MICROSITE_CONFIGURATION": {
        "wl": {
                        "cybersource_config_key": "wl",
                        "domain_prefix": "wl",
                        "university": "wl",
                        "platform_name": "White Label Site",
                        "css_overrides_file": "css/main.css",
                        "SITE_NAME": "wl.exampl.ecom",
                        "course_org_filter": "wlX",
                        "show_partners": false,
                        "ECOMMERCE_PUBLIC_URL_ROOT":  "http://[sitename].lms.local:8002/"
                }
    },
```

LMS does not support checkout with Otto for themed sites/microsites in master branch, so you'd need to switch to [alternative branch](https://github.com/open-craft/edx-platform/tree/ekolpakov/theme-ecommerce-support) until it is merged to master.

**Generate course catalog entries**
1. Create courses in Studio (documented elsewhere)
2. Verify both main site and WL site have some courses - go to LMS (at edx.lms.local and wl.lms.local) and make sure there are some courses displayed in "Find Courses".
3. Go to http://localhost:8002/courses/ and add new course records to course catalog (documented elsewhere)

**Update host machine hosts file**
Add edx.lms.local and wl.lms.local to `/etc/hosts` (or equivalent) as aliases of 127.0.0.1

**Testing instructions:**
For both sites (wl and edx):
1. Log out from both Otto (localhost:8002) and LMS (localhost:8000 / edx.lms.local / wl.lms.local)
2. Log in to LMS as honor (or create new user)
3. Go to course catalogue
4. Enroll in course (if Verified track is avalilable) and upgrade to "Verified" track or add course to a cart (if "professional" track is available).
5. You'll arrive at checkout page. This page contains one "Checkout" button per available payment processor (Cybersource/Paypal). _Expected result:_ buttons are shown only for processors enabled in SiteConfiguration model for that site.
6. Click "Checkout" - you'll be taken to the payment processor site for payment. _Expected result:_ correct PAYMENT_PROCESSOR_CONFIG entry is used. The simplest would be to design visually different payment pages for different configs (i.e. include different subset of optional fields).
7. Complete payment
8. You should be redirected back to appropriate microsite/main site (all payment processors), and have the course upgraded to Verified track (only Paypal - Cybersource can not connect to VM boxes, unless they are exposed to the world - sandbox to test that is currently provisioning)
